### PR TITLE
fix(web): iOS-Safari user agent spoof detection

### LINKED ISSUE #2267: [Web] New iOS setting ("Request Desktop Website") breaks device detection
**Describe the bug**
With new releases of iOS 13.1+, iOS devices include a new setting seen below for Safari:

request_desktop_sites

Enabling this setting will change iOS Safari's user agent setting to perfectly or near-perfectly mirror the current (concurrent with that version's release) macOS user agent string - resulting in broken device detection.

Enabled:
```
> navigator.userAgent
< "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko)"
```

Disabled:
```
> navigator.userAgent
< "Mozilla/5.0 (iPhone; CPU iPhone OS 13_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148"
```

**To Reproduce**
1. Go to Settings > Safari > Request Desktop Website
2. Enable / disable the option
3. Reload whatever page you want to test.

**Expected behavior**
Since KeymanWeb targets touch functionality on iOS devices, we wish to always use "mobile" (as in, "not desktop") functionality.  KeymanWeb should always display the appropriate touch keyboard.

However, in KeymanWeb's current state, it will instead display the desktop OSK... which is much less useful on tablets than the preferred touch version.  When run embedded within the iOS app, the app will ensure that a touch layout is used - there's an extra fallback setting that 'saves' us here - but iOS styling will be missing.

**Screenshots**

keymanweb.com with default settings on iPad, iOS 13.1+:

iPad - enabled

With the setting enabled on iPhone, same iOS version:

iPhone - enabled

As seen above, the iPhone SE is so small that the OSK is completely hidden behind the system keyboard!

The resulting Keyman app keyboard:

iPad in-app results

---

**iPhone/iPad:**
 - OS: iOS 13.1+
 - Device:  any iPad (by default)
 - Keyman version 12.0+
 - Target application: Keyman (in-app and system keyboard)

**KeymanWeb:**
 - OS: iOS 13.1+
 - Device: any iPhone (if manually set) / any iPad (by default)
 - Browser: Safari
 - KeymanWeb Version: 12.0+
 - URL: any (so, keymanweb.com suffices)

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #2267.

This aims to provide a very directed fix for iOS user-agent spoofing as seen in its recently-released offerings.  By testing against specific versions of macOS, we establish a pattern that can be used to adjust device detection accordingly in future versions is the `TouchEvent` check becomes invalid.  Regarding that check itself... https://stackoverflow.com/questions/56578799/tell-ipados-from-macos-on-the-web

Noting the standard aspect ratio of iPad screens to be 4:3 (https://www.lifewire.com/what-is-ipads-screen-resolution-ips-display-1994325) and that iPhones have slimmer form factors, we rely on this ratio to detect the difference betwen iPhones and iPads.

---

Note that without any additional changes, this also fixes in-app and system keyboard behavior for the iOS app on tablets.

### CODE REVIEW

File: web/source/kmwdevice.ts [hunk 1]
```diff
@@ -70,7 +70,54 @@ namespace com.keyman {
...
+          // (by user setting); a new "Request Desktop Website" setting for Safari will
+          // change the user agent string to match a desktop Mac.
+          let regex = /Intel Mac OS X (10(?:_\d+)+)/i;
+          let results = regex.exec(agent);
+          
+          // Match result:  a version string with components separated by underscores.
+          if(results.length > 1 && results[1]) {
+            // Converts version string to array of numbers in major.minor.build order.
+            // Example:  10_14_6 => [10, 14, 6]
+            let versionComponentStrings = results[1].split('_');
```
Comment (darcywong00) [inline web/source/kmwdevice.ts]: What if you tweak the Version() constructor (version.ts l.21) to use a delimiter as a second parameter?

Then you could do `let version = new utils.Version(results[1], '_');`

**Comment by mcdurdin** [comment]
There's one challenge here -- if a user has requested the desktop site but we use the mobile version of KMW, then there may be a disconnect in layout which the site owner has not intended.

For example on keymanweb.com, we really should match the desktop layout with the desktop keyboard even on iPad, if that's what the user has requested. And vice-versa.

One solution may be to include a `deviceMode` parameter or similar for KeymanWeb:

`deviceMode = DEVICE_MODE.DETECT, DEVICE_MODE.TOUCH, DEVICE_MODE.DESKTOP`?

* `DETECT` would be default and use the heuristics as per this PR
* `TOUCH` would *always* show touch (and hence embedded would use this pathway)
* `DESKTOP` would *always* show desktop

Not sure what we have code-wise that already aligns with this but don't want to block off ways for site owners to use KeymanWeb in ways we may not have considered.

File: web/source/kmwdevice.ts (hunk 1 above)
Comment (jahorton) [reply web/source/kmwdevice.ts]: I suppose I _could_ do that, but then the constructor's signature starts getting complex than it already is; TypeScript doesn't allow constructor overloading.

Then again, I just had a different idea that's simpler... why not just `.replace('_', '.')`?

**Comment by jahorton** [comment]
I wrestled with that a bit... is that something we'd want to add to 12.0 stable, though?  We'll need a cherrypick of this so that the iOS app doesn't break, and that means adding extra features to a "stable" version.

I feel the best default (so the one for a cherrypick) is to always present the touch interface.  Just check the screenshots from the issue for some examples why - the iPhone SE version is probably the strongest case for this, considering the 'desktop' OSK gets completely hidden.

**Comment by mcdurdin** [comment]
Yes, I understand. I just think that ultimately the control should rest with the site owner, not KMW. So if we make an interim fix for 12.0 but add the option for 13.0, that might be the best way forward

**Comment by jahorton** [comment]
Considering that the plan is to cherrypick this once approved to stable, mind if I make an issue for the full 13.0 fix and handle that one at a later time? This PR establishes the interim 12.0 fix as-is; I'd rather keep the 13.0-only part separate for easier :cherries:.

**Comment by mcdurdin** [comment]
> Considering that the plan is to cherrypick this once approved to stable, mind if I make an issue for the full 13.0 fix and handle that one at a later time?

Sounds like a good plan to me.

**Pull request merged**